Redesign linux package manager pages

Thraka opened the pull request:
## Summary

- New all-up page for linux distros ([internal preview page](https://review.docs.microsoft.com/en-us/dotnet/core/install/linux?branch=pr-en-us-18300))
- Linux distro all-up pages
  - [x] Ubuntu ([internal preview page](https://review.docs.microsoft.com/en-us/dotnet/core/install/linux-ubuntu?branch=pr-en-us-18300))
  - [x] CentOS ([internal preview page](https://review.docs.microsoft.com/en-us/dotnet/core/install/linux-centos?branch=pr-en-us-18300))
  - [x] Debian ([internal preview page](https://review.docs.microsoft.com/en-us/dotnet/core/install/linux-debian?branch=pr-en-us-18300))
  - [x] Fedora ([internal preview page](https://review.docs.microsoft.com/en-us/dotnet/core/install/linux-fedora?branch=pr-en-us-18300))
  - [x] OpenSUSE ([internal preview page](https://review.docs.microsoft.com/en-us/dotnet/core/install/linux-opensuse?branch=pr-en-us-18300))
  - [x] Redhat ([internal preview page](https://review.docs.microsoft.com/en-us/dotnet/core/install/linux-rhel?branch=pr-en-us-18300))
  - [x] SLES ([internal preview page](https://review.docs.microsoft.com/en-us/dotnet/core/install/linux-sles?branch=pr-en-us-18300))

Fixes #17925

Thraka committed:
Change formt of linux package managers

Thraka committed:
Add a old versions to table

Thraka committed:
Merge branch 'master' into adegeo-17925-distros

Thraka committed:
Changed name to match redirect;removed redirect

Thraka committed:
Merge branch 'adegeo-17925-distros' of github.com:dotnet/docs into adegeo-17925-distros

Thraka committed:
Fix snap link

Thraka committed:
Add file that fell through the crack

Thraka committed:
Add more spacing between the two tables

Thraka committed:
Fix toc

dagood commented on docs/core/install/linux-package-manager-ubuntu.md:
```diff
@@ -0,0 +1,123 @@
+
+## APT (Advanced Package Tool)
+
+Depending on which version of Ubuntu you're using, these commands are slightly different. The changes are usually to URI paths the command uses. You'll want to modify them to match your version of Ubuntu. These changes are called out with placeholders.
```
I think this is asking way too much of our users--I would call it a regression.

Thraka replied:
I agree. I wish I could use auto replacement some how, but it's a lot of work too to duplicate articles for a simple 5 character replacement in a URL...

dagood replied:
No idea what it takes, but is one of these more reasonable?

![image](https://user-images.githubusercontent.com/12819531/81334719-14ea7980-906c-11ea-9773-914b3d57f749.png)

Thraka replied:
That actually is tougher. Unfortunately we have whitelists of tab names.... yeah I know... so we can't just use what makes sense, we have to file tickets and get names approved, and we would have to do that for every version of every distro.

dagood commented on docs/core/install/linux-package-manager-ubuntu.md:
```diff
@@ -0,0 +1,123 @@
+
+Depending on which version of Ubuntu you're using, these commands are slightly different. The changes are usually to URI paths the command uses. You'll want to modify them to match your version of Ubuntu. These changes are called out with placeholders.
+
+The version of .NET Core in this article is `3.1`. You can replace `3.1` with a different version, such as `2.1` or even an unsupported version like `3.0`. If the command fails, it generally means that version isn't available for your Ubuntu version. If you need an unsupported version, you can try other ways to install such as [Snap](runtime.md#install-with-snap).
```
> unsupported version

I think this could be misunderstood to suggest the *version* is unsupported, but we actually just don't support installing that version through the package manager (for a certain distro version, perhaps).

dagood replied:
I think I actually misunderstood, and it is referring to unsupported versions, but giving advice that also applies to versions that aren't available from the package repos.

dagood commented on docs/core/install/linux-package-manager-ubuntu.md:
```diff
@@ -0,0 +1,123 @@
+
+The version of .NET Core in this article is `3.1`. You can replace `3.1` with a different version, such as `2.1` or even an unsupported version like `3.0`. If the command fails, it generally means that version isn't available for your Ubuntu version. If you need an unsupported version, you can try other ways to install such as [Snap](runtime.md#install-with-snap).
+
+The version of Ubuntu used in this article is `19.10`. You can change the feed URIs used here to a supported version such as `16.04`, or `20.04`. You can try to use an unsupported version, such as `17.04` but you're not guaranteed that it will work. If you need an unsupported version, you can try other ways to install such as [Snap](runtime.md#install-with-snap).
```
Seems like this will quickly go out of date, and a pain to maintain the versions for every distro in this format.

dagood reviewed (commented):
> I'm not sure I totally like this design

I'm a bit further down this line, I think it's too complicated.

> I'm really wondering if we should just abandon the whole "package manager" theme from the articles themselves and each distro seriously needs a dedicated page with instructions on how to install via all methods (package manager, snap, install script, download/install). I think that would really server our customers better than splitting info across multiple articles.

I like this. We can share templates for the parts that do work across distros (e.g. the tar.gz install works the same across all glibc distros, snaps are cross-distro by design) but this allows us to diverge easily and give users a **way** better view IMO of what their possibilities are on a given distro.

dagood commented:
@NikolaMilosavljevic @dleeapho

Thraka committed:
Updated ubuntu to demonstrate all ways

Thraka commented:
@dagood OK I added all ways to install on Ubuntu and I cleaned up a lot of the "replace this with that" type of wording. I built a few tables to give alternatives. What do you think?

Thraka committed:
minor

dagood commented:
This still comes across to me as a significant regression from "I can copy-paste a couple of boxes from the docs and I get it installed" to "I have to carefully read the commands and painstakingly substitute a URL in the middle with another URL (probably copy-pasting it into an editor to get that done), then I get it installed". I'm finding that hard to swallow without an exceptionally strong reason.

You *can* set something up where you choose something from the table and it combines with the other step as code:

> | Ubuntu version |                                                                              |
> |----------------|--------------------------------------------------------------------------------|
> | 20.04          | `microsoft_repo_distro_version=20.04` |
> | 19.10          | `microsoft_repo_distro_version=19.10` |
> | 19.04          | `microsoft_repo_distro_version=19.04` |
> 
> ```bash
> wget "https://packages.microsoft.com/config/ubuntu/$microsoft_repo_distro_version/packages-microsoft-prod.deb" -O packages-microsoft-prod.deb
> sudo dpkg -i packages-microsoft-prod.deb
> ```

or

> ```bash
> add_microsoft_repository() {
>   wget "https://packages.microsoft.com/config/ubuntu/$1/packages-microsoft-prod.deb" -O packages-microsoft-prod.deb
>   sudo dpkg -i packages-microsoft-prod.deb
> }
> ```
> 
> | Ubuntu version |                                                                              |
> |----------------|--------------------------------------------------------------------------------|
> | 20.04          | `add_microsoft_repository 20.04` |
> | 19.10          | `add_microsoft_repository 19.10` |
> | 19.04          | `add_microsoft_repository 19.04` |

Still not great IMO, introduces relatively advanced shell concepts to users who might be using Linux for their first time. But at least the scenario goes back to copy-paste.

---

I still think your suggestion to abandon this type of separation is best:

> > I'm really wondering if we should just abandon the whole "package manager" theme from the articles themselves and each distro seriously needs a dedicated page with instructions on how to install via all methods (package manager, snap, install script, download/install). I think that would really server our customers better than splitting info across multiple articles.
> 
> I like this. We can share templates for the parts that do work across distros (e.g. the tar.gz install works the same across all glibc distros, snaps are cross-distro by design) but this allows us to diverge easily and give users a **way** better view IMO of what their possibilities are on a given distro.

dagood commented:
> I still think your suggestion to abandon this type of separation is best:

I see now you did this, but it's per distro, not per distro version. I conflated the two splits, my bad.

leecow commented:
Other than the overall complexity created providing 4 different ways to install, I like the idea of doing a distro per page. Yet another wrinkle we need to work out is how to talk about .NET as provided natively on some Distros (now in CentOS 8 and Fedora 32). I think that once this happens, we should follow the RHEL model and only provide distro instructions. This will reduce the install instructions down to a single option.

Thraka commented:
@leecow Is the SDK provided on those distros? Or just the runtime?

leecow commented:
Both. There will be latency in updates from our releases, particularly when it's not a security release.


**Fedora 32**
```
Last metadata expiration check: 0:00:01 ago on Wed May 13 01:53:55 2020.
========================================================== Name Exactly Matched: dotnet ===========================================================
dotnet.x86_64 : .NET Core CLI tools and runtime
============================================================== Name Matched: dotnet ===============================================================
dotnet-host.x86_64 : .NET command line launcher
dotnet-sdk-3.1.x86_64 : .NET Core 3.1 Software Development Kit
dotnet-hostfxr-3.1.x86_64 : .NET Core command line host resolver
dotnet-runtime-3.1.x86_64 : NET Core 3.1 runtime
dotnet-templates-3.1.x86_64 : .NET Core 3.1 templates
dotnet-apphost-pack-3.1.x86_64 : Targeting Pack for Microsoft.NETCore.App 3.1
dotnet-targeting-pack-3.1.x86_64 : Targeting Pack for Microsoft.NETCore.App 3.1
dotnet-build-reference-packages.x86_64 : Reference packages needed by the .NET Core SDK build
dotnet-sdk-3.1-source-built-artifacts.x86_64 : Internal package for building .NET Core 3.1 Software Development Kit
```
**centos 8**
```
========================================================== Name Exactly Matched: dotnet ===========================================================
dotnet.x86_64 : .NET Core CLI tools and runtime
============================================================== Name Matched: dotnet ===============================================================
dotnet-host.x86_64 : .NET command line launcher
dotnet-sdk-2.1.x86_64 : .NET Core 2.1 Software Development Kit
dotnet-sdk-3.0.x86_64 : .NET Core 3.0 Software Development Kit
dotnet-sdk-3.1.x86_64 : .NET Core 3.1 Software Development Kit
dotnet-hostfxr-3.0.x86_64 : .NET Core command line host resolver
dotnet-hostfxr-3.1.x86_64 : .NET Core command line host resolver
dotnet-runtime-2.1.x86_64 : NET Core 2.1 runtime
dotnet-runtime-3.0.x86_64 : NET Core 3.0 runtime
dotnet-runtime-3.1.x86_64 : NET Core 3.1 runtime
dotnet-sdk-2.1.5xx.x86_64 : .NET Core 2.1.5xx Software Development Kit
dotnet-host-fxr-2.1.x86_64 : .NET Core command line host resolver
dotnet-templates-3.0.x86_64 : .NET Core 3.0 templates
dotnet-templates-3.1.x86_64 : .NET Core 3.1 templates
dotnet-apphost-pack-3.0.x86_64 : Targeting Pack for Microsoft.NETCore.App 3.0
dotnet-apphost-pack-3.1.x86_64 : Targeting Pack for Microsoft.NETCore.App 3.1
dotnet-targeting-pack-3.0.x86_64 : Targeting Pack for Microsoft.NETCore.App 3.0
dotnet-targeting-pack-3.1.x86_64 : Targeting Pack for Microsoft.NETCore.App 3.1
```

Thraka committed:
merge fix

Thraka committed:
master linux page; ubuntu page

Thraka commented:
@dagood @leecow OK trying something new here. Instead of having a table breakdown, merge all install instructions for APT into their own sections, so there is a specific command the user can run.

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,364 @@
+sudo apt-get install -y gpg
+wget -O - https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor -o microsoft.asc.gpg
+sudo mv microsoft.asc.gpg /etc/apt/trusted.gpg.d/
+wget https://packages.microsoft.com/config/ubuntu/{ubuntu-version}/prod.list
```
I'm not too keen on `{netcore-package}` at the end, but it's probably fine since it's the last line--it's easy to see and leave out of my copy-paste. But `{ubuntu-version}` in the middle of a download line, early in the command list seems much worse.

Thraka replied:
I added some intro text to it. Considering this is troubleshooting content, I think the user will be paying more attention to the wording and intro text in this section versus the install sections which are a bit more "blast away" friendly.

dagood replied:
That's a good point. I think though it would be a little bit nicer to use variable formatting, so the user can set up the vars and still simply copy-paste the script. (To keep it simple, I suppose there's no need to call it out, but I'd definitely take advantage of it if I needed to follow these instructions.)

> - `${netcore_package}`\
This represents the .NET Core package you're installing, such as `20.04`. This is in the `sudo apt-get install` command.

> ```
> ...
> wget https://packages.microsoft.com/config/ubuntu/${ubuntu_version}/prod.list
> ...
> sudo apt-get install ${netcore_package}
> ```

(or `$netcore_package`, not sure which type of delimiting makes it clearer.)

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,364 @@
+
+[!INCLUDE [linux-apt-install-31](includes/linux-apt-install-31.md)]
+
+## APT 19.10 (supported)
```
It's a little alarming how large the page got from this, with repeating text... and it will only grow larger as more Ubuntu versions come out. But it is the most reliable way to do this but keep each section infinitely customizable (e.g. for source-build vs. Msft, or distro-specific workarounds if we miss dotnet-runtime-deps for an extended period again).

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,364 @@
+
+For the best compatibility, choose a long-term release (LTS) version.
+
+| .NET Core | Ubuntu version (LTS) | Ubuntu version |
+|-----------|----------------------|----------------|
+| 2.1 (LTS) | 16.04, 18.04, 20.04  | 19.10          |
```
The "Ubuntu version" column isn't immediately obvious what it represents, and a little weird that it's different from how .NET Core LTS is represented. I'd also slightly favor not having a separate list so that 1. it's easier to find the "next-down" and 2. it's easier to find my version in these lists (naturally doing a pseudo-binary-search with my eyes). Throwing this out there (w/ 5.0 for demonstration):

| .NET Core | Ubuntu version |
|-----------|----------------|
| **2.1 (LTS)** | **16.04 (LTS)**, **18.04 (LTS)**, 19.10, **20.04 (LTS)** |
| **3.1 (LTS)** | **16.04 (LTS)**, **18.04 (LTS)**, 19.10, **20.04 (LTS)** |
| 5.0 | **16.04 (LTS)**, **18.04 (LTS)**, 19.10, **20.04 (LTS)** |

Thraka replied:
Fixed

dagood replied:
I'm still not a fan of the out-of-orderness, but other than that looks better now to me.

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,364 @@
+
+The following is a list of currently supported .NET Core releases and the versions of Ubuntu they're supported on. These versions remain supported until either the version of [.NET Core reaches end-of-support](https://dotnet.microsoft.com/platform/support/policy/dotnet-core) or the version of [Ubuntu reaches end-of-life](https://wiki.ubuntu.com/Releases).
+
+For the best compatibility, choose a long-term release (LTS) version.
```
LTS of .NET Core or Ubuntu? This is valid advice for both, but a little hazy.

Thraka replied:
Fixed

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,364 @@
+| 2.1 (LTS) | 16.04, 18.04, 20.04  | 19.10          |
+| 3.1 (LTS) | 16.04, 18.04, 20.04  | 19.10          |
+
+.NET Core versions no longer supported and the Ubuntu LTS releases they were supported with:
```
(Nit) Valid sentence, but I think it should be more direct (easier to read) and possibly say why we're listing them. (Say explicitly that downloads will continue to work?) I think the table should also make it obvious in itself that it's describing non-supported versions, since it's easy (IMO) to skip from table to table and miss the text.

Thraka replied:
Fixed

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,364 @@
+
+.NET Core versions no longer supported and the Ubuntu LTS releases they were supported with:
+
+| .NET Core | Ubuntu version (LTS) |
+|-----------|----------------------|
+| 3.0       | 16.04, 18.04         |
```
Why doesn't this include non-LTS Ubuntu versions?

Thraka replied:
I didn't think a non-supported ubuntu version coupled with a non-supported .NET Core would be as relevant. I can get that information though and add it if you think it's useful.

leecow replied:
I like the idea of only including Ubuntu LTS releases in the 'previous .NET versions' table.

dagood replied:
I'm a little averse to making it hard to find this info, but I agree, it's not as relevant.

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,364 @@
+| .NET Core | Ubuntu version (LTS) | Ubuntu version |
+|-----------|----------------------|----------------|
+| 2.1 (LTS) | 16.04, 18.04, 20.04  | 19.10          |
+| 3.1 (LTS) | 16.04, 18.04, 20.04  | 19.10          |
```
As a user I'd like to be able to click these... the section list is a litty gnarly at my resolution:

> ![image](https://user-images.githubusercontent.com/12819531/82252451-a2ec1d00-9914-11ea-943b-ac73eab487fe.png)

Thraka replied:
Yes. I was planning on linking to the relevant section once I know how they will look.

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,364 @@
+
+[!INCLUDE [linux-apt-install-31](includes/linux-apt-install-31.md)]
+
+## APT 19.04 (not supported)
```
I don't think these stand out quite well enough:

* APT 20.04 (supported)
* APT 19.04 (not supported)

Could we only specify not supported? Use an icon?

Thraka replied:
Check now. What do you think? I removed the APT from the start also. But I think I need to add some intro text to each section though specifying the "APT package manager".

dagood replied:
I like it! I agree some intro text would be good, should probably also call out what the check or cross means (potentially point to the table).

Thraka committed:
Adjust article

Thraka committed:
Fix version

Thraka committed:
Remove bad text

leecow commented on docs/core/install/runtime.md:
````diff
@@ -54,6 +54,32 @@ You can install the .NET Core Runtime with many of the common Linux package mana
+After snap has installed .NET Core SDK, enable the `dotnet` command with the `snap alias` command.
+
+```bash
+sudo snap alias dotnet-sdk.dotnet dotnet
````
should be `sudo snap alias dotnet-runtime-31.dotnet dotnet`

Thraka committed:
Merge branch 'master' of github.com:dotnet/docs into adegeo-17925-distros

Thraka committed:
Add some includes

Thraka committed:
Update the design on ubuntu pages

Thraka committed:
Copy deps to ubuntu

Thraka committed:
add next steps to ubuntu

Thraka committed:
Move notsupported text

Thraka commented:
@dagood @leecow hopefully the last of the design changes.

dagood commented on docs/core/install/includes/linux-ubuntu-apt-prep-intro.md:
```diff
@@ -0,0 +1,11 @@
+
+Installing with APT (Advanced Package Tool) can be done with a few commands. Before installing .NET, you'll need to:
+
+- Add the Microsoft package signing key to the list of trusted keys.
+- Add the repository to the package manager.
+- Install required dependencies.
+
+> [!TIP]
+> Registering the key only needs to be done once per machine.
+
+Open a terminal and run the following commands:
```
This can be misinterpreted as sending you off to do those three things before proceeding. (We've got direct feedback from one person who saw it this way, but I don't think we did anything at that time to address it.) I think adding the tip will make it worse: it visually breaks it up even more, and since it's pointing at one of the three above steps and telling you something about it, it implies adding the key is something you have to actually do. But in Ubuntu, the package repo setup package does that for you.

Is the preamble necessary? Can we be more specific that we're describing what the command does, if we do need to describe it?

(The tip also adds the touchy word "register" back in that we removed earlier.)

Thraka replied:
OK, I reworded this section, can you review it again?

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,336 @@
+
+A snap is a bundle of an app and its dependencies that works without modification across many different Linux distributions. Snaps are discoverable and installable from the Snap Store. For more information about Snap, see [Getting started with Snap](https://snapcraft.io/docs/getting-started).
+
+Only supported versions of .NET Core are available through Snap. Use the following commands to install.
```
I think the rest of the doc has commands just after "following commands" so I expect that, but here it's after a few more paragraphs--which I guess is valid, but not sure if it's intentional.

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,336 @@
+
+Both .NET Core SDK and .NET Core Runtime can be manually installed after they've been downloaded. First, download a binary release for either the SDK or the runtime from one of the following sites:
+
+- [.NET 5.0 preview downloads](https://dotnet.microsoft.com/download/dotnet/5.0)
+- [.NET Core 3.1 downloads](https://dotnet.microsoft.com/download/dotnet-core/3.1)
+- [.NET Core 3.0 downloads](https://dotnet.microsoft.com/download/dotnet-core/3.0)
+- [.NET Core 2.2 downloads](https://dotnet.microsoft.com/download/dotnet-core/2.2)
+- [.NET Core 2.1 downloads](https://dotnet.microsoft.com/download/dotnet-core/2.1)
```
Use a table with LTS and EOL markings like the other sections?

dagood commented on docs/core/install/linux-ubuntu.md:
````diff
@@ -0,0 +1,336 @@
+To extract the runtime and make the .NET Core CLI commands available at the terminal, first download a .NET Core binary release. Then, open a terminal and run the following commands from the directory where the file was saved.
+
+```bash
+mkdir -p $HOME/dotnet && tar zxf aspnetcore-runtime-3.1.0-linux-x64.tar.gz -C $HOME/dotnet
````
Consider quoting to handle spaces:

```suggestion
mkdir -p "$HOME/dotnet" && tar zxf aspnetcore-runtime-3.1.0-linux-x64.tar.gz -C "$HOME/dotnet"
```

(Not necessary for the exports, just the command.)

Thraka committed:
adjust ubuntu; new centos

Thraka committed:
Add fedora/centos/includes

Thraka committed:
feedback

Thraka commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,336 @@
+
+## Dependencies
+
+Ubuntu distributions require the following libraries to be installed:
```
Good points. I'm moving these below the Snap section and I reworded them.

Thraka commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,336 @@
+
+For the best compatibility, choose a long-term release (LTS) version of both Ubuntu and .NET Core.
+
+| .NET Core   | Ubuntu version                                           |
+|-------------|----------------------------------------------------------|
+| 2.1 (LTS)   | **[16.04 (LTS)](#1604-)**, **[18.04 (LTS)](#1804-)**, **[20.04 (LTS)](#2004-)**, [19.10](#1910-) |
+| 3.1 (LTS)   | **[16.04 (LTS)](#1604-)**, **[18.04 (LTS)](#1804-)**, **[20.04 (LTS)](#2004-)**, [19.10](#1910-) |
+| 5.0 Preview | **[16.04 (LTS)](#1604-)**, **[18.04 (LTS)](#1804-)**, **[20.04 (LTS)](#2004-)**, [19.10](#1910-) |
```
Good points. I must have missed the original comment.

Thraka committed:
More feedback|toc

Thraka commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,336 @@
+|-----------|----------------------|
+| 3.0       | 16.04, 18.04         |
+| 2.2       | 16.04, 18.04         |
+| 2.0       | 14.04, 16.04, 18.04  |
```
I'm not seeing it. They are the same order? 🍡

dagood replied:
Here (unsupported): 3.0 => 2.2 => 2.0
Supported: 2.1 => 3.1 => 5.0
Snap: 3.1 => 2.1 => 5.0 and 3.1 => 3.0 => 2.2 => 2.1
Manual install: 5.0 => 3.1 => 3.0 => 2.2 => 2.1

Thraka commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,336 @@
+Ubuntu distributions require the following libraries to be installed:
+
+- liblttng-ust0
+- libcurl3 (for 14.x and 16.x)
```
I don't understand this comment? Are you just agreeing with this format? 😄

dagood replied:
I'm saying 14.x doesn't make sense on first glance to me, and "Ubuntu 14.x" would clarify. I'm also pointing out that this section will need to have numbers like "3.1", "5.0" in the future so it'd get pretty difficult to interpret without clarification.

Thraka replied:
Gotcha! This is the only information I have to go on, if someone can fill out the dependencies information by version that would be great!

Thraka replied:
@dagood just a ping on this message in case you didn't notice it.

dagood replied:
I don't know about the future of these--I did notice more recently though that what you have is the same as what's live at https://docs.microsoft.com/en-us/dotnet/core/install/dependencies?tabs=netcore31&pivots=os-linux#linux-distribution-dependencies, so no need to block on this. (All I'm suggesting is adding "Ubuntu " before each potentially ambiguous version, but if it already exists this way, I'm fine leaving it.)

Thraka committed:
Fix fedora bookmark

Thraka committed:
Standardized support tables

Thraka commented:
OK I think I addressed all of your great suggestions. I added CentOS and Fedora. All three distros are using the same format 😄 I'll get the rest done tomorrow.

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,232 @@
+|-------------|----------------------------------------------------------|
+| 2.1 (LTS)   | [16.04 (LTS) ✔️](#1604-), [18.04 (LTS) ✔️](#1804-), [19.10 ❌](#1910-), [20.04 (LTS) ✔️](#2004-) |
+| 3.1 (LTS)   | [16.04 (LTS) ✔️](#1604-), [18.04 (LTS) ✔️](#1804-), [19.10 ❌](#1910-), [20.04 (LTS) ✔️](#2004-) |
+| 5.0 Preview | [16.04 (LTS) ✔️](#1604-), [18.04 (LTS) ✔️](#1804-), [19.10 ❌](#1910-), [20.04 (LTS) ✔️](#2004-) |
```
Why ❌? I think it *might* be ok to omit a ✔️ to potentially drive usage towards LTS Ubuntus, but putting it on the same terms as lack of support doesn't seem right. Imagine an ❌ next to 20.10 as the rightmost entry for example.

Thraka replied:
Whoops, yes, ❌ shouldn't be there.

dagood commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -0,0 +1,232 @@
+|-----------|----------------------|
+| 3.0       | 16.04, 18.04         |
+| 2.2       | 16.04, 18.04         |
+| 2.0       | 14.04, 16.04, 18.04  |
```
To be more concrete, I think the sections are just missing for these.

Thraka replied:
I originally didn't include 14.04 because I didn't have that documentation actually created. I can't really verify it easily either. However, since it was never documented and is currently out of support I say we just get rid of that.

dagood replied:
That sounds good to me, since it's already not an exhaustive list.

Thraka replied:
Now that I've been playing with it more and thinking about it. I wonder if we should reverse the tables and indicate which versions of .NET were supported and not the other way around.

---

The following table represents the support status of Ubuntu and .NET Core.

- A ✔️ indicates that the version of Ubuntu or .NET Core is still supported.
- A ❌ indicates that the version of Ubuntu or .NET Core is no longer supported.
- When both a version of Ubuntu and a version of .NET Core both have ✔️ that OS and .NET combination are supported.

| Ubuntu | .NET Core 2.x | .NET Core 3.x | .NET 5 Preview |
|-|-|-|-|
| ✔️ [20.04 (LTS)](#2004-)| ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |
| ✔️ [19.10](#1910-)| ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |
| ❌ [19.04](#1904-)| ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |
| ❌ [18.10](#1604-)| ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |
| ✔️ [18.04 (LTS)](#1804-)| ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |
| ❌ [17.10](#1804-)| ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |
| ❌ [17.04](#1804-)| ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |
| ❌ [16.10](#1804-)| ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |
| ✔️ [16.04 (LTS)](#1804-) | ❌ 2.2<br>✔️ 2.1<br>❌ 2.0 | ✔️ 3.1<br>❌ 3.0 | ✔️ 5.0 Preview |

dagood replied:
Showing distro versions as primary makes sense, but I'm not sure about the version columns... seems fairly noisy, for something that I don't think many people will be actively looking for. (I did like having this in a separate table, which seems a little harder to do with distro versions primary.)

Taking it from the top, I think the main entry point is this button:

> ![image](https://user-images.githubusercontent.com/12819531/83803122-a4864680-a671-11ea-858d-5954b54dfcb8.png)

or the tutorial which doesn't mention a version:

> ![image](https://user-images.githubusercontent.com/12819531/83803145-b1a33580-a671-11ea-9ff7-a8e1ac59f35d.png)

I expect that'll take me to https://review.docs.microsoft.com/en-us/dotnet/core/install/linux?branch=pr-en-us-18300. So maybe focusing on how it looks on this page is good? Trimming out the EOL .NET versions removes quite a bit of clutter (also updated availability a bit):

| Ubuntu | .NET Core 2.1 | .NET Core 3.1 | .NET 5 Preview |
|-|-|-|-|
| ✔️ [20.04 (LTS)](#2004-)| ✔️ 2.1 | ✔️ 3.1 | ✔️ 5.0 Preview |
| ✔️ [19.10](#1910-)| ✔️ 2.1 | ✔️ 3.1 | ✔️ 5.0 Preview |
| ❌ [19.04](#1904-)| ✔️ 2.1 | ✔️ 3.1 | ❌ 5.0 Preview |
| ❌ [18.10](#1604-)| ✔️ 2.1 | ❌ 3.1 | ❌ 5.0 Preview |
| ✔️ [18.04 (LTS)](#1804-)| ✔️ 2.1 | ✔️ 3.1 | ✔️ 5.0 Preview |
| ❌ [17.10](#1804-)| ✔️ 2.1 | ❌ 3.1 | ❌ 5.0 Preview |
| ❌ [17.04](#1804-)| ✔️ 2.1 | ❌ 3.1 | ❌ 5.0 Preview |
| ❌ [16.10](#1804-)| ❌ 2.1 | ❌ 3.1 | ❌ 5.0 Preview |
| ✔️ [16.04 (LTS)](#1804-) | ✔️ 2.1 | ✔️ 3.1 | ✔️ 5.0 Preview |

(I think it makes sense for anyone going after EOL .NET versions to have to click through to find more details--I think they'd reasonably be a more advanced user if they have a specific EOL version in mind.)

But maybe we shouldn't even have a table here? The instructions these would link to only describe how to install one version anyway (`dotnet-sdk-3.1`) so we aren't really saying how they could get 2.1 if they want it.

Thraka committed:
Add snap ssl error note

dagood reviewed (approved):
I think the Ubuntu page is looking good--tweaking the tables to be best possible is good, but wanted to say that other than that discussion (and even before) I think this works pretty well.

Thraka committed:
Updates to tables

Thraka committed:
Fix table headers

Thraka committed:
Finished; bug hunt time!

Thraka committed:
Fix links;wording

Thraka committed:
Fix remaining links

Thraka commented:
@NikolaMilosavljevic @dleeapho @dagood Completed! Last chance for any feedback. I will file another issue to track updating the other articles such as install sdk/runtime where we had Linux sections. Currently I zeroed out those sections and pointed them to the root linux article here.

Thraka commented:
@mairaw I believe this change will affect the website. You would now want to link to the `install/linux` article.

dagood commented on docs/core/install/linux.md:
```diff
@@ -0,0 +1,143 @@
+
+## Fedora
+
+Fedora uses DNF as it's package manager.
```
```suggestion
Fedora uses DNF as its package manager.
```

dagood commented on docs/core/install/linux.md:
```diff
@@ -0,0 +1,143 @@
+
+The following table is a list of currently supported .NET Core releases on both CentOS 7 and CentOS 8. These versions remain supported until either the version of [.NET Core reaches end-of-support](https://dotnet.microsoft.com/platform/support/policy/dotnet-core) or the version of CentOS is no longer supported.
+
+| CentOS                   | .NET Core 2.1 | .NET Core 3.1 | .NET 5 Preview |
+|--------------------------|---------------|---------------|----------------|
+| ✔️ [8](linux-centos.md#centos-8-) | ✔️ 2.1        | ✔️ 3.1        | ✔️ 5.0 Preview |
```
I was using 5.0 as a stand-in to look at how these formats scale for the future, you can't actually get them from the package repos--you can install it manually (so this page I guess is still accurate?) but I think I posted some misleading suggestions, sorry. I think these tables should probably only show 2.1 and 3.1 for simplicity?

dagood reviewed (commented):
I don't quite have time to look through it all today, a few initial things I saw though:

IEvangelist commented on docs/core/install/linux-centos.md:
```diff
@@ -0,0 +1,84 @@
+
+| CentOS                   | .NET Core 2.1 | .NET Core 3.1 | .NET 5 Preview |
+|--------------------------|---------------|---------------|----------------|
+| ✔️ [8](#centos-8-) | ✔️ 2.1        | ✔️ 3.1        | ✔️ 5.0 Preview |
```
The ✔️ and ❌ do not work well with screen readers. I think they're okay occasionally, but certainly not in tables. Imagine relying on a screen reader to read a table like this, and those icons not being read - instead of being informative, the table tells them nothing. This is an accessibility concern.

Thraka replied:
We had a discussion on this (maybe 7 months ago? I don't know if it changed? ) But the screen readers we tested did read these correctly. I personally just tried narrator and edge read aloud and both read them fine. Which is great because back then narrator didn't read them correctly and I think we sent them feedback.

IEvangelist commented on docs/core/install/linux-centos.md:
```diff
@@ -0,0 +1,84 @@
+
+[!INCLUDE [package-manager-switcher](./includes/package-manager-heading-hack-pkgname.md)]
+
+## CentOS 8 ✔️
```
Also, not sure I like icons in the headings.

Thraka replied:
Me neither, but some headings were really long and introduced wrapping on smaller screens if we but `(supported)` or `(not supported)` it looked terrible.

IEvangelist commented on docs/core/install/linux-debian.md:
```diff
@@ -0,0 +1,143 @@
+ms.date: 06/04/2020
+---
+
+# Install .NET Core SDK and .NET Core Runtime on Debian
```
If you install the SDK, aren't you also installing that runtime?

IEvangelist commented on docs/core/install/linux-debian.md:
```diff
@@ -0,0 +1,143 @@
+
+# Install .NET Core SDK and .NET Core Runtime on Debian
+
+.NET Core is supported on Debian. This article describes how to install .NET Core on Debian. When a Debian version falls out of support, .NET Core is no longer supported with that version. However, these instructions may help you to get .NET Core running on those versions, even though it isn't supported.
```
```suggestion
This article describes how to install .NET Core on Debian. When a Debian version falls out of support, .NET Core is no longer supported with that version. However, these instructions may help you to get .NET Core running on those versions, even though it isn't supported.
```

IEvangelist commented on docs/core/install/linux-debian.md:
```diff
@@ -0,0 +1,143 @@
+
+## How to install other versions
+
+[!INCLUDE [package-manager-switcher](./includes/package-manager-heading-hack-pkgname.md)]
```
```suggestion
[!INCLUDE [hack-pkgname](./includes/package-manager-heading-hack-pkgname.md)]
```

IEvangelist commented on docs/core/install/linux-debian.md:
```diff
@@ -0,0 +1,143 @@
+
+[!INCLUDE [linux-install-snap](includes/linux-install-snap.md)]
+
+## Dependencies
```
```suggestion
```

Remove duplicate heading

IEvangelist commented on docs/core/install/includes/linux-install-30-dnf.md:
```diff
@@ -0,0 +1,22 @@
+
```
I'm a little concerned about the number of redundancies with these includes. The only real differences between most of them is the version being installed or the install tool. Would it make more sense to have one top-level doc with the structure and leading paragraphs, with some sort of tabs or zone pivot mechanism?

IEvangelist commented on docs/core/install/includes/linux-install-dependencies.md:
```diff
@@ -0,0 +1,22 @@
+
+When you install with a package manager, these libraries are installed for you. But, if you manually install .NET Core or you publish a self-contained app, you'll need to make sure these libraries are installed:
+
+- lttng-ust
```
You should include links to what these libraries are, so that someone reading this can quickly spot-check to ensure they're ok with these additional dependent libraries being installed too.

Thraka replied:
Good suggestion. However, I'm not sure what to link to. This list was taken from the release notes of .NET Core.

dagood replied:
I'd expect Linux users interested in this to know how to use their package manager to find more info:

```
$ apt info liblttng-ust0
Package: liblttng-ust0
Version: 2.11.0-1
Priority: extra
Section: libs
Source: ust
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Jon Bernard <jbernard@debian.org>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 802 kB
Depends: libc6 (>= 2.17), libnuma1 (>= 2.0.11), liburcu6 (>= 0.11.1), liblttng-ust-ctl4 (= 2.11.0-1)
Conflicts: liblttng-ust2
Replaces: liblttng-ust2
Homepage: https://lttng.org/
Download-Size: 160 kB
APT-Sources: http://archive.ubuntu.com/ubuntu focal/main amd64 Packages
Description: LTTng 2.0 Userspace Tracer (tracing libraries)
```

Not opposed to adding more info here, but not sure it's necessary.

IEvangelist commented on docs/core/install/includes/linux-install-dependencies.md:
```diff
@@ -0,0 +1,22 @@
+- libunwind
+- libuuid
+
+If your OpenSSL's version >= 1.1, you'll need to install **compat-openssl10**.
```
```suggestion
If the target runtime environment's OpenSSL version is 1.1 or newer, you'll need to install **compat-openssl10**.
```

IEvangelist commented on docs/core/install/includes/linux-install-dependencies.md:
```diff
@@ -0,0 +1,22 @@
+
+For .NET Core apps that use the *System.Drawing.Common* assembly, you'll also need the following dependency:
+
+- libgdiplus (version 6.0.1 or later)
```
Please include link to library

IEvangelist commented on docs/core/install/includes/linux-install-manual.md:
```diff
@@ -0,0 +1,33 @@
+
+Both .NET Core SDK and .NET Core Runtime can be manually installed after they've been downloaded. First, download a binary release for either the SDK or the runtime from one of the following sites:
```
If you install the SDK, do you need to install the Runtime? I think we should be more explicit about when you need to do one versus the other. For example, the Runtime is usually installed on target environments where the SDK is not needed. The SDK is usually installed by the developers, and it includes the Runtime.

IEvangelist commented on docs/core/install/includes/linux-install-manual.md:
```diff
@@ -0,0 +1,33 @@
+> You can edit your shell profile to permanently add the commands. There are a number of different shells available for Linux and each has a different profile. For example:
+>
+> - **Bash Shell**: *~/.bash_profile*, *~/.bashrc*
+> - **Korn Shell**: *~/.kshrc* or *.profile*
```
```suggestion
> - **KoЯn Shell**: *~/.kshrc* or *.profile*
```

Obviously, do not batch this one - it is meant to be funny! 🤘

IEvangelist commented on docs/core/install/includes/linux-not-supported-ubuntu.md:
```diff
@@ -0,0 +1,2 @@
+
```
I think it would be amazing if DocFx, supported parameter based markdown templating. So that include files could optionally take parameters, from their parent doc. This could really simplify things and remove redundancy - "steps off soapbox".

Thraka replied:
Yeah I know. I think it's been requested for years.

IEvangelist commented on docs/core/install/includes/package-manager-failed-to-fetch-deb.md:
```diff
@@ -1,2 +1,2 @@
 
-While installing the .NET Core package, you may see an error similar to `Failed to fetch ... File has unexpected size ... Mirror sync in progress?`. Generally speaking, this error means that the package feed for .NET Core is being upgraded with newer package versions, and that you should try again later. During an upgrade, the package feed should not be unavailable for more than 30 minutes. If yo ...
+While installing the .NET Core package, you may see an error similar to `Failed to fetch ... File has unexpected size ... Mirror sync in progress?`. Generally speaking, this error means that the package feed for .NET Core is being upgraded with newer package versions, and that you should try again later. During an upgrade, the package feed shouldn't be unavailable for more than 30 minutes. If you ...
```
```suggestion
While installing the .NET Core package, you may see an error similar to `Failed to fetch ... File has unexpected size ... Mirror sync in progress?`. This error means that the package feed for .NET Core is being upgraded with newer package versions, and that you should try again later. During an upgrade, the package feed shouldn't be unavailable for more than 30 minutes. If you continually receive this error for more than 30 minutes, please file an issue at <https://github.com/dotnet/core/issues>.
```

Thraka replied:
I edited your suggestion to add "**could** mean..." as it may not not always be the source of the error.

IEvangelist commented on docs/core/install/sdk.md:
```diff
@@ -48,34 +48,9 @@ export PATH=$PATH:$HOME/dotnet
 
 ::: zone pivot="os-linux"
```
Is this pivot still applicable?

Thraka replied:
Not really, but since it was the current experience I want to keep it until I follow up with a PR redesign the entire page as I'm going to break out windows and macos too.

Thraka committed:
Apply suggestions from code review

Co-authored-by: David Pine <david.pine@microsoft.com>
Co-authored-by: Davis Goodin <dagood@users.noreply.github.com>

IEvangelist reviewed (approved):
Looks good to me, :shipit: when you're ready.

Thraka committed:
Feedback

Thraka committed:
Update instructions for deb10
